### PR TITLE
do stricter version check in GraphHopperStorage.loadExisting

### PR DESCRIPTION
**Pull request by karussell**

### CODE REVIEW

File: core/src/main/java/com/graphhopper/routing/util/EncodingManager.java [hunk 1]
```diff
@@ -124,8 +124,6 @@ public static EncodingManager create(EncodedValueFactory evFactory, FlagEncoderF
             throw new IllegalStateException("Cannot load properties to fetch EncodingManager configuration at: "
                     + dir.getLocation());
 
-        // check encoding for compatibility
-        properties.checkVersions(false);
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/EncodingManager.java]: Is done twice, so no need here

File: core/src/main/java/com/graphhopper/routing/util/EncodingManager.java [hunk 2]
```diff
@@ -289,14 +286,7 @@ public EncodingManager build() {
...
             }
             PMap configuration = new PMap(entryVal);
-
-            FlagEncoder fe = factory.createFlagEncoder(entry, configuration);
-
-            if (configuration.has("version") && fe.getVersion() != configuration.getInt("version", -1))
-                throw new IllegalArgumentException("Encoder " + entry + " was used in version "
-                        + configuration.getLong("version", -1) + ", but current version is " + fe.getVersion());
-
-            resultEncoders.add(fe);
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/EncodingManager.java]: do this check in BaseGraph.loadExisting like we do for encoded values.

File: reader-osm/src/test/java/com/graphhopper/reader/osm/GraphHopperOSMTest.java [hunk 3]
```diff
@@ -475,6 +475,41 @@ public void testFailsForWrongConfig() throws IOException {
...
+        } catch (Exception ex) {
+            assertTrue(ex.getMessage(), ex.getMessage().startsWith("Encoded values do not match"));
+        }
+
+        // different version for car should fail
+        instance = new GraphHopperOSM().setEncodingManager(EncodingManager.create(new FootFlagEncoder(), new CarFlagEncoder() {
+            @Override
+            public int getVersion() {
+                return 0;
+            }
```
Comment (karussell) [inline reader-osm/src/test/java/com/graphhopper/reader/osm/GraphHopperOSMTest.java]: This replaces the test `testEncoderWithWrongVersionIsRejected`

File: core/src/main/java/com/graphhopper/storage/GraphHopperStorage.java [hunk 4]
```diff
@@ -258,6 +258,14 @@ public boolean loadExisting() {
                         + "\nChange configuration to match the graph or delete " + dir.getLocation());
             }
 
+            String encodedValueStr = properties.get("graph.encoded_values");
+            if (!encodedValueStr.isEmpty() && !encodingManager.toEncodedValuesAsString().equalsIgnoreCase(encodedValueStr)) {
+                throw new IllegalStateException("Encoded values do not match:"
+                        + "\nGraphhopper config: " + encodingManager.toEncodedValuesAsString()
+                        + "\nGraph: " + encodedValueStr
+                        + "\nChange configuration to match the graph or delete " + dir.getLocation());
+            }
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/storage/GraphHopperStorage.java]: This is the actual bug fix. See below for the test

**Comment by easbar** [comment]
Not really related here probably, but this looks like a `@Test` annotation is missing (and if I add it the test fails).

https://github.com/graphhopper/graphhopper/blob/2dd229b6e3e7d424691f5d2018c8c70b3fef8445/core/src/test/java/com/graphhopper/routing/util/EncodingManagerTest.java#L199-L203

File: reader-osm/src/test/java/com/graphhopper/reader/osm/GraphHopperOSMTest.java [hunk 5]
```diff
@@ -475,6 +475,74 @@ public void testFailsForWrongConfig() throws IOException {
...
+    public void testFailsForWrongEVConfig() {
+        instance = new GraphHopperOSM().init(
+                new CmdArgs().
+                        put("datareader.file", testOsm3).
+                        put("datareader.dataaccess", "RAM").
+                        put("graph.flag_encoders", "foot,car").
+                        put(Parameters.CH.PREPARE + "weightings", "no")).
+                setGraphHopperLocation(ghLoc);
+        instance.importOrLoad();
+        // older versions do not store this property, ensure that we fail to load it
```
Comment (easbar) [inline reader-osm/src/test/java/com/graphhopper/reader/osm/GraphHopperOSMTest.java]: maybe add something like 'older versions <= 0.12' or reference #1664 here ?

**Pull request merged**